mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java

```diff
@@ -0,0 +1,281 @@
+        return Arrays.stream(extraFields)
+            .filter(field -> field.getHeaderId().equals(InternalDateExtraField.ID_AO))
+            .map(extraField -> ((InternalDateExtraField) extraField).getDateValue())
+            .findFirst()
+            .flatMap(Function.identity());
+    }
+
+    private Optional<Flags> getFlags(ZipEntry entry) throws ZipException {
+        return getStringExtraField(FlagsExtraField.ID_AP, entry).map(Flags::new);
+    }
+
+    private String getMailboxName(ZipEntry current) {
+        return current.getName().substring(0, current.getName().length() - 1);
+    }
+
+    private MailArchiveEntry fromMailboxEntry(ZipEntryWithContent current, Optional<ZipEntry> nextZipEntry) throws ZipException {
+        ZipEntry entry = current.getEntry();
+        if (isLastEntryOrHasNoAnnotationsDirectory(nextZipEntry)) {
+            //no annotation directory or end of iterator => get Mailbox
+            currentMailBox = Optional.empty();
+            return new MailboxWithAnnotationsArchiveEntry(getMailboxName(entry), getMailBoxId(entry).get(), NO_ANNOTATION);
+        } else {
+            // annotation directory => save current mailbox and fetch next entry
+            currentMailBox = Optional.of(new MailboxWithAnnotationsArchiveEntry(getMailboxName(entry), getMailBoxId(entry).get(), NO_ANNOTATION));
+            return next();
+        }
+    }
+
+    private boolean isLastEntryOrNextEntryIsNotOfType(Optional<ZipEntry> nextZipEntry, ZipEntryType mailboxAnnotationDir) throws ZipException {
+        return !nextZipEntry.isPresent() || !getEntryType(nextZipEntry.get()).equals(Optional.of(mailboxAnnotationDir));
+    }
+
+    private boolean isLastEntryOrHasNoAnnotationsDirectory(Optional<ZipEntry> nextZipEntry) throws ZipException {
+        return isLastEntryOrNextEntryIsNotOfType(nextZipEntry, ZipEntryType.MAILBOX_ANNOTATION_DIR);
+    }
+
+    private boolean isLastEntryOrNextEntryIsNotOfType(Optional<ZipEntry> nextZipEntry) throws ZipException {
+        return isLastEntryOrNextEntryIsNotOfType(nextZipEntry, ZipEntryType.MAILBOX_ANNOTATION);
+    }
+
+    private MailArchiveEntry fromMailboxAnnotationEntry(ZipEntryWithContent current, Optional<ZipEntry> next) throws ZipException {
+        MailboxAnnotation annotation = MailboxAnnotation.newInstance(getAnnotationKey(currentMailBox.get(), current.getEntry()),
+            getAnnotationContent(current).orElse(""));
+        MailboxWithAnnotationsArchiveEntry mailboxUpdated = currentMailBox.get().appendAnnotation(annotation);
+
+        if (isLastEntryOrNextEntryIsNotOfType(next)) {
+            currentMailBox = Optional.empty();
+            return mailboxUpdated;
+        } else {
+            currentMailBox = Optional.of(mailboxUpdated);
+            return next();
+        }
+    }
+
+    private Optional<MailArchiveEntry> fromMessageEntry(ZipEntryWithContent entryWithContent) throws ZipException {
```

Comment:
I would love each `MailArchiveEntry` to have a factory accepting a `Collection<ZipEntryWithContent>` as a parameter

That way we can extract some responsibilities of this too big class.

WDYT?

Reply:
Yes, it seems  good to me